dattri/algorithm/dvemb.py

```diff
@@ -0,0 +1,588 @@
+"""Data Value Embedding (DVEmb) attributor for trajectory-specific data influence."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Iterator
+    from typing import Dict, List, Optional, Tuple
+
+    from torch import Tensor
+    from torch.utils.data import DataLoader
+
+import math
+
+import torch
+from torch import nn
+from torch.func import grad, vmap
+from tqdm import tqdm
+
+from dattri.func.utils import flatten_func, flatten_params
+
+
+class DVEmbAttributor:
+    """Data Value Embedding (DVEmb) attributor.
+
+    DVEmb captures temporal dependence in training by computing data value embeddings
+    that approximate trajectory-specific leave-one-out influence. This implementation
+    stores embeddings for each epoch separately, allowing for epoch-specific analysis.
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,
+        loss_func: Callable,
+        device: str = "cpu",
+        proj_dim: Optional[int] = None,
+        factorization_type: str = "none",
+    ) -> None:
+        """Initializes the DVEmb attributor.
+
+        Args:
+            model: The PyTorch model to be attributed.
+            loss_func: A per-sample loss function. It should take model parameters
+                       and a single data sample (unbatched) as input.
+            device: The device to run computations on (e.g., "cpu", "cuda").
+            proj_dim: The dimension for projection (if used).
+            factorization_type: Type of gradient factorization to use. Options are
+                                "none"(default),
+                                "kronecker"(same with paper),
+                                or "elementwise"(more memory-efficient).
+
+        Raises:
+            ValueError: If an unknown factorization type is provided.
+        """
+        self.model = model
+        self.device = device
+        self.criterion = loss_func
+        self.projector = None
+        self.use_factorization = factorization_type != "none"
+        self.factorization_type = factorization_type
+        self.projection_dim = proj_dim
+
+        if not self.use_factorization:
+
+            @flatten_func(self.model)
+            def vmap_loss_fn(
+                params: Tensor,
+                data_tensors: Tuple[Tensor, Tensor],
+            ) -> Tensor:
+                inputs = data_tensors[0].unsqueeze(0)
+                targets = data_tensors[1].unsqueeze(0)
```

Comment:
If we use AttributionTask (https://github.com/TRAIS-Lab/dattri/blob/main/dattri/task.py#L40) This will be a user defined function and we don't really need to assume user's data is a tuple of 2 tensors.